implementations/rust/ockam/ockam_api/src/ui/terminal/mod.rs

```diff
@@ -401,6 +402,13 @@ impl<W: TerminalWriter + Debug> Terminal<W, ToStdOut> {
             },
         };
 
+        // Log the plain message if present
+        if let Some(plain) = self.mode.output.plain.as_ref() {
+            if msg.ne(plain) {
```

Comment:
Having to compare full messages here is a bit of a red flag. 

I would suggest to refactor some code in that function:
 
- Have a method on `self.mode.output.get_message(format, is_tty): Result<OutputMessage>`.
- That method returns an error if no output has been selected, otherwise it applies the priority rules
- `OutputMessage` is an enum with alternatives `Plain(String) | Machine(String) | Json(String)`.
- The formatting of the `Json` type can be done directly in `get_message` (or raise an error if formatting can't be done).
- Then pattern-matching on `(OutputMessage, self.output.plain.as_ref())` we can decide:
   - To log both messages if `OutputMessage` is `Json` or `Machine` and `self.output.plain` is defined.
   - To log only `OutputMessage` otherwise.

One advantage is that we can now unit test the priority rules.

Reply:
I've added this 👍